04_snake_game/main.go

```diff
@@ -10,97 +10,145 @@ import (
+}
+func newApple(maxX, maxY int) apple {
+	return apple{coord{rand.Intn(maxX), rand.Intn(maxY)}}
 }
 
-// newGame returns a new game state.
-func newGame() game {
-	// Sets game field dimensions to the size of the terminal.
-	w, h := termbox.Size()
+func newGame() game { // newGame returns a new game state.
+	w, h := termbox.Size() // Sets game field dimensions to the size of the terminal.
 	return game{
 		fieldWidth:  w,
 		fieldHeight: h,
 		sn:          newSnake(w, h),
 		v:           coord{1, 0},
+		ap:          newApple(w, h),
 	}
 }
 
 // drawSnakePosition draws the current snake position (as a debugging
 // information) in the buffer.
+func writeGameOver(g game) {
+	gOC := termbox.ColorRed
+	gBc := termbox.ColorDefault
+	str := "Game Over"
+	writeText((g.fieldWidth/2)-len(str), (g.fieldHeight/2)-1, str, gOC, gBc)
+}
 func drawSnakePosition(g game) {
-	str := fmt.Sprintf("(%d, %d)", g.sn.pos.x, g.sn.pos.y)
+	str := fmt.Sprintf("(%d, %d)", g.sn.pos[0].x, g.sn.pos[0].y)
 	writeText(g.fieldWidth-len(str), 0, str, snakeFgColor, snakeBgColor)
 }
-
-// drawSnake draws the snake in the buffer.
-func drawSnake(sn snake) {
-	termbox.SetCell(sn.pos.x, sn.pos.y, snakeBody, snakeFgColor, snakeBgColor)
+func drawScore(g game) {
+	temporaryScore := 10
+	str := fmt.Sprintf("Score: %d", temporaryScore)
+	writeText((g.fieldWidth/2)-len(str), 0, str, snakeFgColor, snakeBgColor)
 }
-
-// Redraws the terminal.
-func draw(g game) {
-	// Clear the old "frame".
-	termbox.Clear(snakeFgColor, snakeBgColor)
+func drawHearts(g game) { // maybe in future
+	heart := "♥"
+	str := fmt.Sprintf("%s%s%s", heart, heart, heart)
+	writeText(0, 0, str, snakeFgColor, snakeBgColor)
+}
+func drawApple(ap apple) {
+	termbox.SetCell(ap.pos.x, ap.pos.y, appleBody, appleFg, appleBg)
+}
+func drawSnake(sn snake) { // drawSnake draws the snake in the buffer.
+	for _, pos := range sn.pos {
+		termbox.SetCell(pos.x, pos.y, snakeBody, snakeFgColor, snakeBgColor)
+	}
+}
+func drawBorder(g game) {
+	ch := rune('#')
+	bColor := termbox.ColorCyan
+	bBackC := termbox.ColorDefault
+	for i := 0; i != g.fieldWidth; i++ {
+		termbox.SetCell(i, 1, ch, bColor, bBackC)
+	}
+	for i := 1; i != g.fieldHeight; i++ {
+		termbox.SetCell(0, i, ch, bColor, bBackC)
+	}
+	for i := g.fieldWidth; i != 0; i-- {
+		termbox.SetCell(i, g.fieldHeight-1, ch, bColor, bBackC)
+	}
+	for i := g.fieldHeight; i != 0; i-- {
+		termbox.SetCell(g.fieldWidth-1, i, ch, bColor, bBackC)
+	}
+}
+func borderExists(g game) {
+	if g.sn.pos[0].x == 1 || g.sn.pos[0].y == 2 || g.sn.pos[0].x == g.fieldWidth-2 || g.sn.pos[0].y == g.fieldHeight-2 {
+		writeGameOver(g)
+		time.Sleep(100 * time.Millisecond)
+		return
+	}
+}
+func draw(g game) { // Redraws the terminal.
+	termbox.Clear(snakeFgColor, snakeBgColor) // Clear the old "frame".
+	drawApple(g.ap)
+	drawBorder(g)
+	writeGameOver(g)
```

Comment:
This function is called at every game step, and "Game over" is always written on the screen. Do you understand why? Can you fix it?